src/smart-account.ts

```diff
@@ -0,0 +1,314 @@
+import {
+    ethers,
+    BlockTag,
+    Signer,
+    TypedDataDomain,
+    TypedDataField,
+    getAddress,
+    SigningKey,
+    BytesLike,
+} from "ethers";
+import { Provider } from "./provider";
+import { EIP712Signer } from "./signer";
+import { Wallet } from "./wallet";
+import { EIP712_TX_TYPE, DEFAULT_GAS_PER_PUBDATA_LIMIT, serializeEip712 } from "./utils";
+import { Address, TransactionResponse, TransactionRequest, TransactionLike, Eip712Meta } from "./types";
+
+/**
+ * The custom signing method is used to sign transactions with a custom logic
+ * @param transaction - transaction to sign
+ * @param defaultSigner - default signer to sign transaction
+ * @param keys - private keys to sign transaction
+ * @returns signed transaction
+ */
+type CustomSigningMethod = (
+    transaction: TransactionRequest,
+    defaultSigner: ethers.Signer,
+    keys: string[] | string | ethers.SigningKey,
+) => Promise<string>;
+
+export class SmartAccount extends EIP712Signer implements Signer {
+    readonly address: string;
+    provider: Provider;
+    readonly customSigningMethod?: CustomSigningMethod;
+    readonly privateKeys: string[] | string | ethers.SigningKey;
+    private defaultSigner: ethers.Signer;
+    // public eip712: EIP712Signer;
+    readonly #signingKey: SigningKey;
+
+    constructor(
+        address: string,
+        privateKeys: string[] | string | ethers.SigningKey,
```

Comment:
Why do we want privateKeys here? I would just pass a Signer or a signing function, like whenever it has to sign a transaction it calls it.

Reply:
I think the signing function is the final optional parameter `signingMethod`, but I think it's because `Wallet` requires a `privateKey` in its constructor. @uF4No is that correct?

Reply:
correct, `signingMethod` is the last and optional param. I'm trying to accommodate for different scenarios like:

- an smart contract account that uses the default signing but has other functionalities like spend limit, or app controls.
- multisig smart contract accounts
- totally custom smart accounts with it's own signing method

In the first case, the class takes care of it as it's the default signing method. For the other two scenarios, users will need to pass the key(s) and signingMethod. I think this is the option that can cover most different scenarios for different types of signing. The keys are passed to the signingMethod to be used as needed.

I understand @Romsters point tough but, my understanding is that any custom signing method would need its keys to actually sing the transaction, that's why I require them in the constructor and pass them later when calling the custom signing in `singTransaction`